OJ-1984: Set up npm workspaces for lambda repos

CharlesIC opened the pull request:
### Set up npm workspaces

Enable the monorepo config for npm so all packages for all lambdas can be installed at once by running `npm install` at the root of the repo. Other npm commands can be run across the entire repo as well.

The SAM template has been adjusted to account for the fact that the root of the repo is the base dir now when running SAM builds, since it needs to refer to the root `package.json` file. The handler is specified from the root of the repo as well.

npm packages will be installed to the `npm_modules` directory at the root of the repo rather than in individual lambda directories. Common packages used by all lambdas are listed in the root `package.json` file, and each lambda still has its own `package.json` file to install additional packages for that lambda.

**_Note on package isolation between lambdas_**
Since all packages for all lambdas are installed in the same root directory, it is possible for a lambda to refer to a package installed in the directory but not listed in its `package.json` file or the root `package.json` file. This only happens when running code locally. For SAM builds and deployments, package isolation is preserved. This is due to the fact that SAM copies the lambda source code to a temporary directory and installs its dependencies, and it does so in isolation for each lambda (resulting in sizable disk use for caching since all dependencies are installed/duplicated independently for each lambda even if they're exactly the same). Therefore, this change in how packages are installed does not affect deployments.

---

- Add `--enable-source-maps` flag to lambda node options
  This should be enabled when esbuild compiles code with source maps enabled.
- Clean up `package.json` files
  We don't publish npm packages (except for one repo), so we don't need to specify the licence, author, version and main properties.
- Add SAM scripts to the root of the repo

### Testing

Generate a repo with some lambdas by running this command a few times - select repo on the first run, and then lambdas
```
npm run plop -- --dest <your dest>
```

Navigate to `dest` and these commands can now be run at the root
```
npm install
npm run test
npm run sam:build
```

CharlesIC committed:
OJ-1984: Set up npm workspaces for lambda repos

Enable the monorepo config for npm so all packages for all lambdas can be installed at once by running `npm install` at the root of the repo. Other npm commands can be run across the entire repo as well.

npm packages will be installed to the `npm_modules` directory at the root of the repo rather than in individual lambda directories. Common packages used by all lambdas are listed in the root package.json file, and each lambda still has its own package.json file to install additional packages for that lambda.

The SAM template has been adjusted to account for the fact that the root of the repo is the base dir now when running SAM builds, since it needs to refer to the root package.jsonn file. The handler is specified from the root of the repo as well.

Add SAM scripts to the root of the repo.

CharlesIC committed:
BAU: Add `--enable-source-maps` flag to lambda node options

This should be enabled when esbuild compiles code with source maps enabled

merlinc commented on frontend/templates/npm/package.json.hbs:
```diff
@@ -1,8 +1,7 @@
-  "version": "1.0.0",
   "description": "Frontend for the {{ titleCase criName }} Credential Issuer",
-  "main": "index.js",
+  "author": "GOV.UK One Login",
```
Author is a [field for one person](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#people-fields-author-contributors) - we should remove it as per the rest of this PR

merlinc commented on lambda/index.mjs:
```diff
@@ -39,7 +39,7 @@ export default async function (plop) {
           type: "addMany",
           destination: "./",
-          base: "templates/config",
+          base: "templates/repo-config",
```
`repo-config` isn't a correct name for what these templates do.

In the frontend repo I referred to them as `npm` ... which is more accurate but still could do with some improvement.

CharlesIC replied:
It's not just npm config (unlike in frontend) - there's eslint and jest config in there as well. `repo-config` is the best I could come up with. I'm open to suggestions but I don't think `npm` is a good name in this case

merlinc replied:
https://github.com/govuk-one-login/ipv-cri-templates/tree/main/frontend/templates/npm - This is the similar folder in the `frontend` repo. 

Repo config is not accurate as a) there is already set of actions to init a repository  & b) not every repo will need node or npm.

Perhaps `node` is a better name, as that's a little bit wider than npm?

If you're renaming it here, you should also rename it in the frontend repo, so things are consistent.

merlinc commented on lambda/templates/infrastructure/template.hbs:
```diff
@@ -17,10 +17,14 @@ Conditions:
       !If [UsePermissionsBoundary, !Ref PermissionsBoundary, !Ref AWS::NoValue]
+    Environment:
+      Variables:
+        NODE_OPTIONS: --enable-source-maps
```
:raised_hands:

merlinc commented on lambda/templates/repo-config/package.json.hbs:
```diff
@@ -1,23 +1,38 @@
   "name": "{{ kebabCase criName }}",
   "description": "",
+  "author": "GOV.UK One Login",
+  "workspaces": [
```
:+1:

merlinc commented on lambda/templates/repo-config/package.json.hbs:
```diff
@@ -1,23 +1,38 @@
     "test:coverage": "npm run unit -- --coverage",
     "pre-commit": "pre-commit run --all-files --show-diff-on-failure",
-    "pc": "npm run pre-commit"
+    "pc": "npm run pre-commit",
```
There's no dependency for `pre-commit` and we shouldn't need to run this normally in the generated repo.

(I admit, the workflow when running tempaltes + validating correctness is weird, and required a lot of `pre-commit run --all-files --show-diff-on-failure`)

CharlesIC replied:
It is useful to be able to run this at any time - not just when you commit, and it's good to encourage people to use the tools we have configured.
Now, each dev can make this is an alias on their machine so it's not a big deal if it's not included here but I just think it's useful to make it easy for people to run checks on their local branches.

merlinc replied:
If people want to run pre-commit manually at any time, they can do it. We don't need to pollute the package.json with extra runners. If pre-commit changes their API in version 4, we'll have to update every single repo for this shortcut.

As you said, if developers find this useful - and don't want to reuse their history - they can create an alias on their machine.

merlinc commented on lambda/templates/repo-config/package.json.hbs:
```diff
@@ -1,23 +1,38 @@
     "pre-commit": "pre-commit run --all-files --show-diff-on-failure",
-    "pc": "npm run pre-commit"
+    "pc": "npm run pre-commit",
+    "sam:validate": "cd infrastructure && sam validate && sam validate --lint",
```
Can we add this into a linting task in a GHA?

CharlesIC replied:
Yes, we should - I'll add it in the PR that touches the workflows later

merlinc commented on lambda/templates/repo-config/package.json.hbs:
```diff
@@ -1,23 +1,38 @@
+    "sam:build": "cd infrastructure && sam build"
+  },
+  "dependencies": {
+    "@aws-lambda-powertools/commons": "^1.14.2",
```
We're using dependabot, so we should install without the caret ranges - that ensures we're installing the expected version if we ever have to regenerate the lockfile.

CharlesIC replied:
OK, I think it's useful to get minor version updates automatically but I'm not precious about it

merlinc commented on frontend/templates/cucumber/tests/browser/package.json.hbs:
```diff
@@ -1,12 +1,9 @@
 {
   "name": "ipv-cri-{{ kebabCase criName }}-front-tests",
-  "version": "1.0.0",
```
These changes are not part of the "setup npm workspaces for lambda repos" so should be pulled out into a separate PR.

merlinc reviewed (changes requested):
I like a lot of what's in this PR - workspaces looks like where we should have taken this previously.

CharlesIC committed:
BAU: Group frontend config files under `config`

For consistency with the lambdas generator, refer to all the repo-level config files as "config".

CharlesIC committed:
BAU: Remove pre-commit script from generated package.json files

Keep the scripts minimal and let developers set up their own scripts on their machines.
